File: BadgeGenerator.java
// BadgeGenerator.java
import com.google.zxing.BarcodeFormat;
import com.google.zxing.WriterException;
import com.google.zxing.common.BitMatrix;
import com.google.zxing.qrcode.QRCodeWriter;
import com.google.zxing.client.j2se.MatrixToImageConfig;
import com.google.zxing.client.j2se.MatrixToImageWriter;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.*;
import java.nio.file.*;
import javax.imageio.ImageIO;
import com.google.gson.*;
import java.util.*;
import java.util.List;

public class BadgeGenerator {
    static class Template {
        int badgeWidth, badgeHeight;
        String background;
        int nameFontSize, titleFontSize, companyFontSize;
        int[] qrPosition;
        int[] textPosition;
        String textColor;
    }

    private static Template defaultTemplate() {
        Template t = new Template();
        t.badgeWidth = 600;
        t.badgeHeight = 400;
        t.background = "#FFFFFF";
        t.nameFontSize = 40;
        t.titleFontSize = 24;
        t.companyFontSize = 28;
        t.qrPosition = new int[]{380, 80};
        t.textPosition = new int[]{50, 100};
        t.textColor = "#000000";
        return t;
    }

    private static Template loadTemplate(String path) throws IOException {
        Gson gson = new Gson();
        try (Reader reader = Files.newBufferedReader(Paths.get(path))) {
            return gson.fromJson(reader, Template.class);
        }
    }

    private static int hexToRgb(String hex) {
        hex = hex.replace("#", "");
        return Integer.parseInt(hex, 16);
    }

    private static BufferedImage createQR(String data, int size, String color, String logoPath) throws WriterException, IOException {
        QRCodeWriter writer = new QRCodeWriter();
        BitMatrix matrix = writer.encode(data, BarcodeFormat.QR_CODE, size, size);
        int qrColor = hexToRgb(color);
        MatrixToImageConfig config = new MatrixToImageConfig(qrColor, 0xFFFFFFFF);
        BufferedImage qr = MatrixToImageWriter.toBufferedImage(matrix, config);
        if (logoPath != null && !logoPath.isEmpty()) {
            BufferedImage logo = ImageIO.read(new File(logoPath));
            int logoSize = size / 4;
            Image scaled = logo.getScaledInstance(logoSize, logoSize, Image.SCALE_SMOOTH);
            BufferedImage logoImg = new BufferedImage(logoSize, logoSize, BufferedImage.TYPE_INT_ARGB);
            Graphics2D g = logoImg.createGraphics();
            g.drawImage(scaled, 0, 0, null);
            g.dispose();
            int x = (size - logoSize) / 2;
            int y = (size - logoSize) / 2;
            Graphics2D qrG = qr.createGraphics();
            qrG.drawImage(logoImg, x, y, null);
            qrG.dispose();
        }
        return qr;
    }

    public static void generateBadge(String data, String name, String title, String company,
                                     String output, int size, String color, String logoPath,
                                     Template template) throws Exception {
        BufferedImage qr = createQR(data, size, color, logoPath);
        // Create badge
        BufferedImage badge = new BufferedImage(template.badgeWidth, template.badgeHeight, BufferedImage.TYPE_INT_RGB);
        Graphics2D g = badge.createGraphics();
        g.setColor(Color.decode(template.background));
        g.fillRect(0, 0, template.badgeWidth, template.badgeHeight);
        g.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON);
        g.setColor(Color.decode(template.textColor));
        // Draw text
        Font nameFont = new Font("Arial", Font.PLAIN, template.nameFontSize);
        Font titleFont = new Font("Arial", Font.PLAIN, template.titleFontSize);
        Font companyFont = new Font("Arial", Font.PLAIN, template.companyFontSize);
        int tx = template.textPosition[0];
        int ty = template.textPosition[1];
        g.setFont(nameFont);
        g.drawString(name, tx, ty);
        if (title != null && !title.isEmpty()) {
            g.setFont(titleFont);
            g.drawString(title, tx, ty + template.nameFontSize + 10);
        }
        if (company != null && !company.isEmpty()) {
            g.setFont(companyFont);
            g.drawString(company, tx, ty + template.nameFontSize + template.titleFontSize + 20);
        }
        // Draw QR
        int qx = template.qrPosition[0];
        int qy = template.qrPosition[1];
        g.drawImage(qr, qx, qy, null);
        g.dispose();

        // Save
        String ext = output.substring(output.lastIndexOf('.') + 1).toLowerCase();
        if (ext.equals("png")) {
            ImageIO.write(badge, "png", new File(output));
        } else {
            // PDF not implemented for simplicity
            ImageIO.write(badge, "png", new File(output));
        }
        System.out.println("Badge saved to " + output);
    }

    public static void main(String[] args) {
        String data = null, name = null, title = null, company = null, output = "badge.png";
        int size = 200;
        String color = "#000000";
        String logo = null;
        String templatePath = null;
        String batch = null;
        String outputDir = ".";

        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--data": data = args[++i]; break;
                case "--name": name = args[++i]; break;
                case "--title": title = args[++i]; break;
                case "--company": company = args[++i]; break;
                case "--output": output = args[++i]; break;
                case "--size": size = Integer.parseInt(args[++i]); break;
                case "--color": color = args[++i]; break;
                case "--logo": logo = args[++i]; break;
                case "--template": templatePath = args[++i]; break;
                case "--batch": batch = args[++i]; break;
                case "--output-dir": outputDir = args[++i]; break;
            }
        }

        try {
            Template template = (templatePath != null) ? loadTemplate(templatePath) : defaultTemplate();

            if (batch != null) {
                // CSV parsing
                List<String[]> rows = new ArrayList<>();
                try (BufferedReader br = new BufferedReader(new FileReader(batch))) {
                    String line;
                    String[] header = null;
                    while ((line = br.readLine()) != null) {
                        if (header == null) {
                            header = line.split(",");
                        } else {
                            rows.add(line.split(","));
                        }
                    }
                }
                for (String[] row : rows) {
                    Map<String, String> map = new HashMap<>();
                    for (int i = 0; i < header.length; i++) {
                        if (i < row.length) map.put(header[i], row[i]);
                    }
                    String d = map.getOrDefault("data", "");
                    String n = map.getOrDefault("name", d);
                    String t = map.getOrDefault("title", "");
                    String c = map.getOrDefault("company", "");
                    String out = map.getOrDefault("output", n + ".png");
                    String outPath = outputDir + "/" + out;
                    generateBadge(d, n, t, c, outPath, size, color, logo, template);
                }
            } else {
                if (data == null) { System.err.println("Error: --data required"); System.exit(1); }
                if (name == null) name = data;
                generateBadge(data, name, title, company, output, size, color, logo, template);
            }
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }
    }
}

File: BadgeGenerator.kt
// BadgeGenerator.kt
import com.google.zxing.BarcodeFormat
import com.google.zxing.common.BitMatrix
import com.google.zxing.qrcode.QRCodeWriter
import com.google.zxing.client.j2se.MatrixToImageConfig
import com.google.zxing.client.j2se.MatrixToImageWriter
import java.awt.*
import java.awt.image.BufferedImage
import java.io.File
import javax.imageio.ImageIO
import com.google.gson.Gson
import java.nio.file.Files
import java.nio.file.Paths

data class Template(
    val badgeWidth: Int,
    val badgeHeight: Int,
    val background: String,
    val nameFontSize: Int,
    val titleFontSize: Int,
    val companyFontSize: Int,
    val qrPosition: List<Int>,
    val textPosition: List<Int>,
    val textColor: String
)

fun defaultTemplate() = Template(
    badgeWidth = 600,
    badgeHeight = 400,
    background = "#FFFFFF",
    nameFontSize = 40,
    titleFontSize = 24,
    companyFontSize = 28,
    qrPosition = listOf(380, 80),
    textPosition = listOf(50, 100),
    textColor = "#000000"
)

fun loadTemplate(path: String): Template {
    val json = String(Files.readAllBytes(Paths.get(path)))
    return Gson().fromJson(json, Template::class.java)
}

fun hexToRgb(hex: String): Int {
    return hex.replace("#", "").toInt(16)
}

fun createQR(data: String, size: Int, color: String, logoPath: String?): BufferedImage {
    val writer = QRCodeWriter()
    val matrix = writer.encode(data, BarcodeFormat.QR_CODE, size, size)
    val config = MatrixToImageConfig(hexToRgb(color), 0xFFFFFFFF)
    var qr = MatrixToImageWriter.toBufferedImage(matrix, config)
    if (logoPath != null) {
        val logo = ImageIO.read(File(logoPath))
        val logoSize = size / 4
        val scaled = logo.getScaledInstance(logoSize, logoSize, Image.SCALE_SMOOTH)
        val logoImg = BufferedImage(logoSize, logoSize, BufferedImage.TYPE_INT_ARGB)
        val g = logoImg.createGraphics()
        g.drawImage(scaled, 0, 0, null)
        g.dispose()
        val x = (size - logoSize) / 2
        val y = (size - logoSize) / 2
        val g2 = qr.createGraphics()
        g2.drawImage(logoImg, x, y, null)
        g2.dispose()
    }
    return qr
}

fun generateBadge(data: String, name: String, title: String?, company: String?,
                  output: String, size: Int, color: String, logoPath: String?,
                  template: Template) {
    val qr = createQR(data, size, color, logoPath)
    val badge = BufferedImage(template.badgeWidth, template.badgeHeight, BufferedImage.TYPE_INT_RGB)
    val g = badge.createGraphics()
    g.color = Color.decode(template.background)
    g.fillRect(0, 0, template.badgeWidth, template.badgeHeight)
    g.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON)
    g.color = Color.decode(template.textColor)
    val nameFont = Font("Arial", Font.PLAIN, template.nameFontSize)
    val titleFont = Font("Arial", Font.PLAIN, template.titleFontSize)
    val companyFont = Font("Arial", Font.PLAIN, template.companyFontSize)
    val tx = template.textPosition[0]
    val ty = template.textPosition[1]
    g.font = nameFont
    g.drawString(name, tx.toFloat(), ty.toFloat())
    if (!title.isNullOrEmpty()) {
        g.font = titleFont
        g.drawString(title, tx.toFloat(), (ty + template.nameFontSize + 10).toFloat())
    }
    if (!company.isNullOrEmpty()) {
        g.font = companyFont
        g.drawString(company, tx.toFloat(), (ty + template.nameFontSize + template.titleFontSize + 20).toFloat())
    }
    val qx = template.qrPosition[0]
    val qy = template.qrPosition[1]
    g.drawImage(qr, qx, qy, null)
    g.dispose()
    val ext = output.substringAfterLast('.')
    if (ext.lowercase() == "png") {
        ImageIO.write(badge, "png", File(output))
    } else {
        ImageIO.write(badge, "png", File(output)) // fallback
    }
    println("Badge saved to $output")
}

fun main(args: Array<String>) {
    var data: String? = null
    var name: String? = null
    var title: String? = null
    var company: String? = null
    var output = "badge.png"
    var size = 200
    var color = "#000000"
    var logoPath: String? = null
    var templatePath: String? = null
    var batchPath: String? = null
    var outputDir = "."

    var i = 0
    while (i < args.size) {
        when (args[i]) {
            "--data" -> data = args[++i]
            "--name" -> name = args[++i]
            "--title" -> title = args[++i]
            "--company" -> company = args[++i]
            "--output" -> output = args[++i]
            "--size" -> size = args[++i].toInt()
            "--color" -> color = args[++i]
            "--logo" -> logoPath = args[++i]
            "--template" -> templatePath = args[++i]
            "--batch" -> batchPath = args[++i]
            "--output-dir" -> outputDir = args[++i]
        }
        i++
    }

    val template = if (templatePath != null) loadTemplate(templatePath) else defaultTemplate()

    if (batchPath != null) {
        // CSV parsing (simple split)
        val lines = File(batchPath).readLines()
        val header = lines.first().split(",")
        for (line in lines.drop(1)) {
            val row = line.split(",")
            val map = header.zip(row).toMap()
            val d = map["data"] ?: ""
            val n = map["name"] ?: d
            val t = map["title"] ?: ""
            val c = map["company"] ?: ""
            val out = map["output"] ?: "$n.png"
            val outPath = "$outputDir/$out"
            generateBadge(d, n, t, c, outPath, size, color, logoPath, template)
        }
    } else {
        if (data == null) { System.err.println("Error: --data required"); System.exit(1) }
        if (name == null) name = data
        generateBadge(data!!, name!!, title, company, output, size, color, logoPath, template)
    }
}
